panic when using nil *Time pointer

```go
package main

import (
	"github.com/bytedance/sonic"
	"time"
)

func main() {
	a := struct {
		T *time.Time
	}{}
	sonic.Marshal(&a)
}
```

panic msg:

```bash
panic: value method time.Time.MarshalJSON called using nil *Time pointer

goroutine 1 [running]:
time.(*Time).MarshalJSON(0x0, 0x0, 0x4286e7c500000000, 0xc0001cdd80, 0x0, 0x0)
	<autogenerated>:1 +0x93
github.com/bytedance/sonic/encoder.encodeJsonMarshaler(0xc0000a63a8, 0x2cb1160, 0x0, 0x0, 0x0)
	/Users/tyrionlong/go/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210723083932-f9632ab87317/encoder/primitives.go:82 +0x31
github.com/bytedance/sonic/encoder.encodeTypedPointer(0xc0000a63a8, 0x125b020, 0xc000010028, 0xc000300000, 0x0, 0xc000165ed0)
	/Users/tyrionlong/go/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210723083932-f9632ab87317/encoder/primitives.go:77 +0x163
github.com/bytedance/sonic/encoder.EncodeInto(0xc0000a63a8, 0x125b020, 0xc0000be1d0, 0xc000200000, 0xc000165f30)
	/Users/tyrionlong/go/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210723083932-f9632ab87317/encoder/encoder.go:67 +0xc5
github.com/bytedance/sonic/encoder.Encode(0x125b020, 0xc0000be1d0, 0x1019701, 0x0, 0x0, 0x125d560, 0xc00008c058)
	/Users/tyrionlong/go/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210723083932-f9632ab87317/encoder/encoder.go:47 +0xbb
github.com/bytedance/sonic.Marshal(...)
	/Users/tyrionlong/go/pkg/mod/github.com/bytedance/sonic@v0.0.0-20210723083932-f9632ab87317/sonic.go:30
main.main()
	/Users/tyrionlong/Downloads/foo/main.go:12 +0x3d
exit status 2
```

--- a/encoder/compiler.go
+++ b/encoder/compiler.go
@@ -416,7 +416,7 @@ func (self *_Compiler) compileRec(p *_Program, sp int, vt reflect.Type, pv bool)
 
     /* check for `json.Marshaler` */
     if vt.Implements(jsonMarshalerType) {
-        p.rtt(_OP_marshal, vt)
+        self.compileMarshaler(p, _OP_marshal, vt, jsonMarshalerType)
         return
     }
 
@@ -428,7 +428,7 @@ func (self *_Compiler) compileRec(p *_Program, sp int, vt reflect.Type, pv bool)
 
     /* check for `encoding.TextMarshaler` */
     if vt.Implements(encodingTextMarshalerType) {
-        p.rtt(_OP_marshal_text, vt)
+        self.compileMarshaler(p, _OP_marshal_text, vt, encodingTextMarshalerType)
         return
     }
 
@@ -819,4 +819,24 @@ func (self *_Compiler) compileInterface(p *_Program, vt reflect.Type) {
     p.pin(x)
     p.add(_OP_null)
     p.pin(e)
+}
+
+func (self *_Compiler) compileMarshaler(p *_Program, op _Op, vt reflect.Type, mt reflect.Type) {
+    pc := p.pc()
+    vk := vt.Kind()
+
+    /* direct receiver */
+    if vk != reflect.Ptr || !vt.Elem().Implements(mt) {
+        p.rtt(op, vt)
+        return
+    }
+
+    /* value receiver with a pointer type, check for nil before calling the marshaler */
+    p.add(_OP_is_nil)
+    p.rtt(op, vt)
+    i := p.pc()
+    p.add(_OP_goto)
+    p.pin(pc)
+    p.add(_OP_null)
+    p.pin(i)
 }

--- a/issue58_test.go
+++ b/issue58_test.go
@@ -1,0 +1,41 @@
+/*
+ * Copyright 2021 ByteDance Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package sonic
+
+import (
+    `testing`
+
+    `github.com/stretchr/testify/require`
+)
+
+type (
+    Issue58ValueReceiver   struct {}
+    Issue58PointerReceiver struct {}
+)
+
+func (_ Issue58ValueReceiver) MarshalJSON() ([]byte, error)    { return []byte(`"value"`), nil }
+func (_ *Issue58PointerReceiver) MarshalJSON() ([]byte, error) { return []byte(`"pointer"`), nil }
+
+func TestIssue58_NilPointerOnValueMethod(t *testing.T) {
+    v := struct {
+        X *Issue58ValueReceiver
+        Y *Issue58PointerReceiver
+    }{}
+    buf, err := Marshal(v)
+    require.NoError(t, err)
+    require.Equal(t, []byte(`{"X":null,"Y":"pointer"}`), buf)
+}

